### PR TITLE
[RzIL] Rework Variable Handling

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

There are now three kinds of vars: global, local and local pure. Global
and local pure are exactly like in BAP, local ones are defined by their
initial set op and have the scope of a single lifted instruction.

The set op now handles both global and local vars, let is now pure and
binds only inside its body. Vars have static types, global and local are
always mutable, local pure naturally immutable.

The var op specifies the kind of variable to take from, and the
individual variable sets are separate. This corresponds to BAP's
behavior where the kind of variable is part of the identifier.

Variable content storage has also been rewritten and RzILBag removed.

**Test plan**

see the added unit tests

### CODE REVIEW

File: librz/core/cil.c [hunk 1]
```diff
@@ -574,35 +539,39 @@ RZ_IPI void rz_core_analysis_rzil_vm_status(RzCore *core, const char *var_name,
...
-			rz_warn_if_reached();
-			break;
-		}
-		if (var_name) {
-			break;
-		}
-		if (rz_strbuf_length(p_sb(p.ptr)) > 95) {
-			rz_cons_printf("%s\n", rz_strbuf_get(p_sb(p.ptr)));
-			rz_strbuf_fini(p_sb(p.ptr));
+	RzPVector *global_vars = rz_il_vm_get_all_vars(rzil->vm, RZ_IL_VAR_KIND_GLOBAL);
```
Comment (XVilka) [inline librz/core/cil.c:542]: I think it worth extracting this vars initialization into a separate function. Could be reused somewhere else separately, as well.

File: librz/il/definitions/value.c [hunk 2]
```diff
@@ -83,18 +108,24 @@ RZ_API void rz_il_value_free(RZ_NULLABLE RzILVal *val) {
...
 		break;
-	case RZ_IL_VAR_TYPE_UNK:
 	default:
 		break;
 	}
-
 	free(val);
 }
+
+RZ_API RzILSortPure rz_il_value_get_sort(RZ_NONNULL RzILVal *val) {
```
Comment (XVilka) [inline librz/il/definitions/value.c:127]: Add Doxygen

File: librz/il/definitions/variable.c [hunk 3]
```diff
@@ -36,6 +33,118 @@ RZ_API void rz_il_variable_free(RZ_NULLABLE RzILVar *var) {
...
+	free(kv->key);
+	rz_il_variable_free(kv->value);
+}
+
+static void val_ht_free(HtPPKv *kv) {
+	free(kv->key);
+	rz_il_value_free(kv->value);
+}
+
+RZ_API bool rz_il_var_set_init(RzILVarSet *vs) {
```
Comment (XVilka) [inline librz/il/definitions/variable.c:60]: Doxygen, please

File: librz/il/definitions/variable.c [hunk 4]
```diff
@@ -36,6 +33,118 @@ RZ_API void rz_il_variable_free(RZ_NULLABLE RzILVar *var) {
...
+	if (!kv) {
+		return NULL;
+	}
+	RzILVal *r = kv->value;
+	kv->value = NULL;
+	ht_pp_delete(vs->contents, name);
+	return r;
+}
+
+RZ_API bool rz_il_var_set_bind(RzILVarSet *vs, const char *name, RZ_OWN RzILVal *val) {
```
Comment (XVilka) [inline librz/il/definitions/variable.c:129]: Doxygen

File: librz/il/definitions/variable.c [hunk 5]
```diff
@@ -36,6 +33,118 @@ RZ_API void rz_il_variable_free(RZ_NULLABLE RzILVar *var) {
...
+			RZ_LOG_ERROR("Attempted to bind mis-sorted value to variable \"%s\"\n", name);
+		}
+		rz_il_value_free(val);
+		return false;
+	}
+	ht_pp_update(vs->contents, name, val);
+	return true;
+}
+
+RZ_API RZ_BORROW RzILVar *rz_il_var_set_get(RzILVarSet *vs, const char *name) {
```
Comment (XVilka) [inline librz/il/definitions/variable.c:148]: Same, etc

File: librz/include/rz_il/definitions/sort.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#define RZ_IL_SORT_H
+
+#include <rz_types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \brief Type and Sort identifiers for values in the IL
```
Comment (XVilka) [inline librz/include/rz_il/definitions/sort.h:15]: You could use `\file` and grouping with `\@{`

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cil.c (hunk 1 above)
Comment (thestr4ng3r) [reply librz/core/cil.c:542]: I'm not sure I understand what you mean. This is not initializing anything, it is only printing the var contents.

**Review by XVilka — APPROVED** [review]

**Pull request merged**